### ISSUE TITLE
Cellpose memory usage and runtimes

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal#108]
Together with @mfranzon, we explored the memory usage of cellpose for image labeling.

## TL;DR
* The limiting factor for current image labeling is the standard RAM, not the GPU memory.
* The large memory usage is correctly part of cellpose, and not due to our (dask) implementation.
* Labeling a `(10,2160,2560)` reaches a peak of approx 18 G of RES memory, meaning that on a 64 G node we cannot hope to run more than 3 FOVs simultaneously (and we should be very careful when number of Z planes increases). This agrees with previous knowledge in https://github.com/fractal-analytics-platform/fractal/issues/64#issuecomment-1177514846.
* Conjecture: an important part of the memory-usage peak could be related to `flows`, large arrays computed by cellpose and that we currently do not use. Would it be possible to not compute them? (the question is for the cellpose devs)
* Here and there in this issue one can find typical runtimes (in the plots or in the logs).

## Context
* The tests below are performed via SLURM (parsl is not involved).
* Code corresponds to e76dddee67944bd2f43107277ec3129cb7dbab94.
* RES memory is traced via Python's memory_profiler.
* VIRT memory is traced by hand, by simply looking at `htop` in real time. Note that the VIRT data seem to match with the ones traced by parsl (in independent tests). We don't report data for VIRT memory here, but it typically corresponds to a constant shift with respect to RES (sometimes around 10 G).
* We always 3D-segment a single channel of a 2x2 well, with 10 Z planes.
* We explicitly cap the number of threads used by dask's `map_blocks`, to see the scaling with parallel execution.
* We **do not include relabeling**.
* We do not modify the default `batch_size=8` of cellpose's `eval`.
* The submission script is this one:
```
#!/bin/bas 
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=8
#SBATCH --partition=gpu
#SBATCH --time=10:00:00
date
echo
NUM_THREADS=$1
rm -r /data/active/fractal/tests/Temporary_data_UZH_1_well_2x2_sites_singlefov/20200812-CardiomyocyteDifferentiation14-Cycle1.zarr/B/03/0/labels
poetry run mprof run --multiprocess --include-children image_labeling.py -z /data/active/fractal/tests/Temporary_data_UZH_1_well_2x2_sites_singlefov/20200812-CardiomyocyteDifferentiation14-Cycle1.zarr/B/03/0/ --chl_list "A01_C01" -lc "A01_C01" --num_threads $NUM_THREADS
mv LOG_image_labeling LOG_image_labeling_${NUM_THREADS}_threads
```
* Figures are created with commands like `mprof plot -t "num_threads=1" -f -o 'figure_n1.png'`.
* Just calling `core.use_gpu` (from cellpose) once looks like an innocent action, but it produces a bump of approx 2 G in RES memory. Still, it is needed.

## Memory usage
Here are the memory-usage traces for `num_threads=1` and `num_threads=2`.

RES memory usage, `num_threads=1`, where we see four blocks of single-site executions.

![figure_n1](https://user-images.githubusercontent.com/3862206/179006051-e93b92e8-0b2f-4f2a-a50d-9405e1c7f0b8.png)

RES memory usage, `num_threads=2`, where we see two blocks of two-sites executions.

![figure_n2](https://user-images.githubusercontent.com/3862206/179008869-5e2b3beb-28b4-4b55-84d5-fd84c30543e6.png)

Note that the large peaks at the end of each block are probably (partly) due to `flows`. For instance we know that:
```
mask.shape: (10, 2160, 2560)
flows[1].shape: (3, 10, 2160, 2560)
```
Is it possible to avoid computing these flows in cellpose?

## Logs

Here are relevant logs for the `num_threads=1` and `num_threads=2` cases (note the asynchronous execution in the second case):
```
Start image_labeling task for /data/active/fractal/tests/Temporary_data_UZH_1_well_2x2_sites_singlefov/20200812-CardiomyocyteDifferentiation14-Cycle1.zarr/B/03/0/
relabeling: False
Total well shape/chunks:
(10, 4320, 5120)
((1, 1, 1, 1, 1, 1, 1, 1, 1, 1), (2160, 2160), (2560, 2560))

After map_block, mask will have (10, 4320, 5120) and chunks ((1, 1, 1, 1, 1, 1, 1, 1, 1, 1), (2160, 2160), (2560, 2560))

[(0, 0, 0)] START Cellpose | column: <class 'numpy.ndarray'>, (10, 2160, 2560) | do_3D: True
[(0, 0, 0)] END   Cellpose | Elapsed: 214.5070 seconds | mask shape: (10, 2160, 2560), mask dtype: uint16 (before recast to <class 'numpy.uint32'>), max(mask): 2203
[(0, 0, 1)] START Cellpose | column: <class 'numpy.ndarray'>, (10, 2160, 2560) | do_3D: True
[(0, 0, 1)] END   Cellpose | Elapsed: 215.5441 seconds | mask shape: (10, 2160, 2560), mask dtype: uint16 (before recast to <class 'numpy.uint32'>), max(mask): 1764
[(0, 1, 0)] START Cellpose | column: <class 'numpy.ndarray'>, (10, 2160, 2560) | do_3D: True
[(0, 1, 0)] END   Cellpose | Elapsed: 220.7558 seconds | mask shape: (10, 2160, 2560), mask dtype: uint16 (before recast to <class 'numpy.uint32'>), max(mask): 2211
[(0, 1, 1)] START Cellpose | column: <class 'numpy.ndarray'>, (10, 2160, 2560) | do_3D: True
[(0, 1, 1)] END   Cellpose | Elapsed: 201.0475 seconds | mask shape: (10, 2160, 2560), mask dtype: uint16 (before recast to <class 'numpy.uint32'>), max(mask): 1937
```

```
Start image_labeling task for /data/active/fractal/tests/Temporary_data_UZH_1_well_2x2_sites_singlefov/20200812-CardiomyocyteDifferentiation14-Cycle1.zarr/B/03/0/
relabeling: False
Total well shape/chunks:
(10, 4320, 5120)
((1, 1, 1, 1, 1, 1, 1, 1, 1, 1), (2160, 2160), (2560, 2560))

After map_block, mask will have (10, 4320, 5120) and chunks ((1, 1, 1, 1, 1, 1, 1, 1, 1, 1), (2160, 2160), (2560, 2560))

[(0, 0, 0)] START Cellpose | column: <class 'numpy.ndarray'>, (10, 2160, 2560) | do_3D: True
[(0, 0, 1)] START Cellpose | column: <class 'numpy.ndarray'>, (10, 2160, 2560) | do_3D: True
[(0, 0, 0)] END   Cellpose | Elapsed: 278.7268 seconds | mask shape: (10, 2160, 2560), mask dtype: uint16 (before recast to <class 'numpy.uint32'>), max(mask): 2203
[(0, 1, 0)] START Cellpose | column: <class 'numpy.ndarray'>, (10, 2160, 2560) | do_3D: True
[(0, 0, 1)] END   Cellpose | Elapsed: 281.4100 seconds | mask shape: (10, 2160, 2560), mask dtype: uint16 (before recast to <class 'numpy.uint32'>), max(mask): 1764
[(0, 1, 1)] START Cellpose | column: <class 'numpy.ndarray'>, (10, 2160, 2560) | do_3D: True
[(0, 1, 1)] END   Cellpose | Elapsed: 261.7807 seconds | mask shape: (10, 2160, 2560), mask dtype: uint16 (before recast to <class 'numpy.uint32'>), max(mask): 1937
[(0, 1, 0)] END   Cellpose | Elapsed: 281.7634 seconds | mask shape: (10, 2160, 2560), mask dtype: uint16 (before recast to <class 'numpy.uint32'>), max(mask): 2211
```

## Possibly related (or maybe not)
* https://github.com/MouseLand/cellpose/issues/178
* https://github.com/MouseLand/cellpose/issues/171
* https://cellpose.readthedocs.io/en/latest/api.html#cellpose.models.Cellpose.eval
* https://github.com/MouseLand/cellpose/issues/534

**Comment by tcompa** [comment]
Another quick comment:
Setting `batch_size=16` (twice as large as the default) in cellpose's core.eval leads to an increased memory usage on the GPU (roughly from 3 to 6 G). At the moment it's unclear whether this also speeds up the calculations significantly (to be verified).

```
batch_size (int (optional, default 8)) – number of 224x224 patches to run simultaneously on the GPU (can make smaller or bigger depending on GPU memory usage)
```

**Comment by tcompa** [comment]
Possibly related: "Memory management for parallel processing" (https://github.com/MouseLand/cellpose/issues/207).

**Comment by jluethi** [comment]
Great documentation of the process @tcompa !

As discussed at a call today:
- Flows are a core part of cellpose, I don't think we get around those
- In general, optimizing cellpose will be out of scope for our project. But very good to know the limiting factor is cellpose, not something in our parsl setup
- Curious for the results of doing relabeling on the 9x8 well test set after cellpose has run :) The relabeling can be made lazy later to scale even further, but it already handles decently large datasets :)

**Comment by tcompa** [comment]
Here are some info on a successful **labeling+relabeling on a 9x8 well with 19 Z planes**.

This is for a run submitted via SLURM (no parsl), and with `num_threads=2` simultaneous cellpose executions. Relabeling is included. Total runtime is 3h20m, and relabeling is only taking ~90 seconds. The memory trace is under control throughout labeling (note that there are now 19 Z planes, while earlier tests in this issue were for 10 planes), and relabeling doesn't play any role in memory usage:

![figure_72_FOVs_labeling_real_relabeling_true](https://user-images.githubusercontent.com/3862206/179199718-f20add0b-9c7e-423a-9dbd-d6b0f1439520.png)

LOGS: [LOG_72_FOVs_labeling_real_relabeling_true_num_threads_2.txt](https://github.com/fractal-analytics-platform/fractal/files/9119806/LOG_72_FOVs_labeling_real_relabeling_true_num_threads_2.txt)

Note: the total number of labels is >100k, so that a `uint16` mask would not be sufficient (and indeed we are using `uint32` by default).

***

As a reference, we also report info on a run were the cellpose run was replaced by a `np.ones_like(column)` trivial operation. This shows a huge enhancement of memory usage for relabeling, which we attribute to the fact that there are no zeros in the mask (which are likely treated in an optimized way in zarr arrays).

![figure_72_FOVs_labeling_fake_relabeling_true](https://user-images.githubusercontent.com/3862206/179200526-3518f3fe-ac28-4f2e-82df-8307cfa9b35e.png)

LOGS: [LOG_72_FOVs_labeling_fake_relabeling_true_num_threads_1.txt](https://github.com/fractal-analytics-platform/fractal/files/9119827/LOG_72_FOVs_labeling_fake_relabeling_true_num_threads_1.txt)
